### PR TITLE
[REF] html_editor: clean the resource API for plugins

### PR DESCRIPTION
**Pull request by Goaman**
This commit change the resource API for the plugins in 3 ways.

# 1) The disposal of resources

Instead of

```js
class MyPlugin extends Plugin {
    /** @type { (p: MyPlugin) => Record<string, any> } */
    static resources = (p) => ({
        my_handlers: p.handler.bind(p)
    })
}

```

We now have
```js
class MyPlugin extends Plugin {
    resources = {
        my_handlers: this.handler.bind(this)
    }
}
```

# 2) Resource getter
To access a specific resource within a plugin, instead of

```js
this.resources['my_handler']
```

We now have
```js
this.getResource('my_handler')
```

If no other plugin disposed the specified resource, the getter will return an empty array.

# 3) Sequences for all resources
This commit give the ability to specify a sequence for any resource item, removing the need for the consumer of the resource to have to setup it.

Instead of
```js
class MyPluginA extends Plugin {
    resources = {
        my_handlers: {handler: this.handler.bind(this), sequence: 100}
        my_data: {id: 'some-id', sequence: 100}
    }
}
class MyPluginB extends Plugin {
    constructor() {
        this.handlers = this.getResource('my_handlers')
            .sort((a, b) => a.sequence - b.sequence)
    }
    myMethod() {
        for (const {handler} of this.myHandlers) {
            handler();
        }
    }
}
```

We now have
```js
class MyPluginA extends Plugin {
    resources = {
        my_handlers: withSequence(100, this.handler.bind(this)),
        my_data: withSequence(100, {id: 'some-id'}),
    }
}
class MyPluginB extends Plugin {
    myMethod() {
        for (const handler of this.getResource('my_handlers')) {
            handler();
        }
    }
}
```

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/static/src/project_sharing/editor/project_sharing_media_plugin.js [hunk 1]
```diff
@@ -3,12 +3,11 @@ import { MediaPlugin } from "@html_editor/main/media/media_plugin";
...
-        const resources = MediaPlugin.resources(p);
-        delete resources.toolbarCategory;
-        delete resources.toolbarItems;
-        return resources;
-    };
+    constructor(...args) {
+        super(...args);
+        delete this.resources.toolbarCategory;
+        delete this.resources.toolbarItems;
+    }
```
Comment (FrancoisGe) [inline addons/project/static/src/project_sharing/editor/project_sharing_media_plugin.js:10]: ...

File: addons/html_editor/static/src/core/format_plugin.js [hunk 2]
```diff
@@ -100,12 +100,12 @@ export class FormatPlugin extends Plugin {
...
                 icon: "fa-eraser",
                 title: _t("Remove Format"),
-                hasFormat: hasFormat(p),
+                hasFormat: hasFormat(this),
             },
         ],
         arrows_should_skip: (ev, char, lastSkipped) => char === "\u200b",
-        onBeforeInput: { handler: p.onBeforeInput.bind(p), sequence: 60 },
-    });
+        onBeforeInput: withSequence(20, this.onBeforeInput.bind(this)),
```
Comment (Jinjiu96) [inline addons/html_editor/static/src/core/format_plugin.js:107]: It looks like a big number change, might have problems with other plugins' `onBeforeInput` :thinking:

**Review by fdardenne — APPROVED** [review]
@robodoo r+

**Comment by fdardenne** [comment]
@robodoo r+

**Comment by fdardenne** [comment]
@robodoo r+

**Pull request closed without merging**